wallet: add tracepoints and algorithm information to coin selection

achow101 opened the pull request:
Tracepoints can be useful for coin selection as they would allow us to observe what is being selected, selection parameters, and calculation results. So this PR adds 4 new tracepoints:

1. After `SelectCoins` returns in order to observe the `SelectionResult`
2. After the first `CreateTransactionInternal` to observe the created transaction
3. Prior to the second `CreateTransactionInternal` to notify that the optimistic avoid partial spends selection is occurring
4. After the second `CreateTransactionInternal` to observe the created transaction and inform which solution is being used.

This PR also adds an algorithm enum to `SelectionResult` so that the first tracepoint will be able to report which algorithm was used to produce that result.

The primary use case for these tracepoints is in running coin selection simulations. The script I use to run these simulations use these tracepoints in order to gather data on the algorithm used and the calculated waste.

achow101 committed:
wallet: track which coin selection algorithm produced a SelectionResult

achow101 committed:
wallet: Add some tracepoints for coin selection

josibake commented:
Concept ACK

S3RK commented:
Huge Concept ACK!

Looking at existing tracing contexts, it seems `wallet` context would be better level of granularity. Also `doc/tracing.md` needs to be updated to mention new context.

vincenzopalazzo commented on src/wallet/coinselection.h:
```diff
@@ -200,21 +200,32 @@ struct OutputGroup
  */
 [[nodiscard]] CAmount GetSelectionWaste(const std::set<CInputCoin>& inputs, CAmount change_cost, CAmount target, bool use_effective_value = true);
 
+enum SelectionAlgorithm : uint8_t
+{
+    BNB = 0,
+    KNAPSACK = 1,
+    SRD = 2,
+    MANUAL = 3,
+};
+
```
There is any motivation to use the plain `enum` here and not the `enum class`?

Maybe I'm missing some things but usually, the enum class is preferred in C++.

achow101 replied:
Done

fanquake commented:
cc @0xB10C

0xB10C commented:
Cool, Concept ACK! Will review.

We might want to start adding tracepoint interface tests (see https://github.com/bitcoin/bitcoin/pull/24358) in the same PR when adding new tracepoints. This makes review a lot easier and provides an usage example.

achow101 committed:
wallet: compute waste for SelectionResults of preset inputs

When we use only manually specified inputs, we should still calculate
the waste so that if anything later on calls GetWaste (in order to log
it), there won't be an error.

vincenzopalazzo reviewed (approved):
ACK https://github.com/bitcoin/bitcoin/commit/64338c1a1f4dadb371919042f6be4399ec16b8ec

jb55 commented:
Concept ACK

jonatack commented:
Concept ACK

0xB10C commented:
fwiw: I think to move this forward at least documentation in [doc/tracing.md#tracepoint-documentation](https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md#tracepoint-documentation) and interface tests similar to #24358 are needed.

I'm not too familiar with coin selection and don't know what e.g. the `coin_selection:aps_create_tx_internal` tracepoint traces. This makes reviewing cumbersome for me.

achow101 committed:
doc: document coin selection tracepoints

achow101 committed:
test: Add test for coinselection tracepoints

achow101 commented:
I've added docs and a test.

laanwj commented:
Concept ACK, thanks for adding tracepoints. I'll try to have a more detailed look at this soon.

0xB10C commented on test/functional/interface_usdt_coinselection.py:
```diff
@@ -0,0 +1,210 @@
+     See https://github.com/bitcoin/bitcoin/blob/master/doc/tracing.md#context-coin_selection
+"""
+
+import ctypes
+from io import BytesIO
```
CI linter complains:
```
test/functional/interface_usdt_coinselection.py:10:1: F401 'ctypes' imported but unused
test/functional/interface_usdt_coinselection.py:11:1: F401 'io.BytesIO' imported but unused
```

0xB10C commented on test/functional/interface_usdt_coinselection.py:
```diff
@@ -0,0 +1,210 @@
+
+    def get_tracepoints(self, expected_types):
+        events = []
+        try:
+            for i in range(0, 6):
+                event = self.bpf["coin_selection_events"].pop()
+                assert_equal(event.wallet_name.decode(), self.default_wallet_name)
+                assert_equal(event.type, expected_types[i])
+                events.append(event)
+            else:
+                # If the loop exits successfully instead of throwing a KeyError, then we have had
+                # more events than expected. There should be no more than 5 events.
+                assert False
+        except KeyError:
+            assert_equal(len(events), len(expected_types))
+            return events
```
I found this to be unclear, probably because I was unfamiliar with Python's `for/else`. If I understand correctly, you want to check that exactly `len(expected_types)` tracepoints were triggered? 

```suggestion
        try:
            for i in range(0, len(expected_types) + 1):
                event = self.bpf["coin_selection_events"].pop()
                assert_equal(event.wallet_name.decode(), self.default_wallet_name)
                assert_equal(event.type, expected_types[i])
                events.append(event)
            else:
                # If the loop exits successfully instead of throwing a KeyError, then we have had
                # more events than expected. There should be no more than `len(expected_types)` events.
                assert False
        except KeyError:
            assert_equal(len(events), len(expected_types))
            return events
```

achow101 replied:
Done

0xB10C commented on test/functional/interface_usdt_coinselection.py:
```diff
@@ -0,0 +1,210 @@
+    s64 selected_value;
+
+    // create tx event
+    u8 success;
```
```suggestion
    bool success;
```

achow101 replied:
Done

0xB10C commented on test/functional/interface_usdt_coinselection.py:
```diff
@@ -0,0 +1,210 @@
+    s32 change_pos;
+
+    // aps create tx event
+    u8 use_aps;
```
```suggestion
    bool use_aps;
```

0xB10C commented on test/functional/interface_usdt_coinselection.py:
```diff
@@ -0,0 +1,210 @@
+                    waste = event.waste
+                sc_events.append(event)
+            elif event.type == 2:
+                success = event.success == 1
```
with event.success being a `bool` rather than a `u8`
```suggestion
                success = event.success
```

achow101 replied:
Done

0xB10C commented on test/functional/interface_usdt_coinselection.py:
```diff
@@ -0,0 +1,210 @@
+                is_aps = True
+            elif event.type == 4:
+                assert is_aps
+                if event.use_aps == 1:
```
with event.use_aps being a `bool` rather than a `u8`

```suggestion
                if event.use_aps:
```

0xB10C commented on test/functional/interface_usdt_coinselection.py:
```diff
@@ -0,0 +1,210 @@
+        wallet.sendtoaddress(wallet.getnewaddress(), 10)
+        events = self.get_tracepoints([1, 2, 3, 1, 4])
+        success, use_aps, algo, waste, change_pos = self.determine_selection_from_usdt(events)
+        assert_equal(success, True)
+        assert_greater_than(change_pos, -1)
```
does it make sense to check `use_aps`, `algo`, and `waste` here too?

glozow replied:
As a sanity check that it's not malformed, you could `assert algo in ["bnb", "srd", "knapsack", "manual"]`

0xB10C commented on test/functional/interface_usdt_coinselection.py:
```diff
@@ -0,0 +1,210 @@
+        events = self.get_tracepoints([1, 2])
+        success, use_aps, algo, waste, change_pos = self.determine_selection_from_usdt(events)
+        assert_equal(success, True)
+        assert_equal(use_aps, None)
```
does it make sense to check `algo`, `change_pos`, and `waste` here too?

0xB10C commented:
The tracepoint part looks good. Left a few comments.

I also wrote a quick and dirty bpftrace script to test the documentation. In case that's helpful to anyone:

<details>
<summary>
bpftrace script
</summary>

``` bpftrace
#!/usr/bin/env bpftrace

BEGIN
{
  printf("Logging coin selection\n")
}

usdt:./src/bitcoind:coin_selection:selected_coins
{
  $name = str(arg0);
  $algo = str(arg1);
  $target = (int64) arg2;
  $waste = (int64) arg3;
  $inputsum = (int64) arg4;
  printf("selected_coins: wallet=%s, algo=%s, target=%d, waste=%d, inputsum=%d\n", $name, $algo, $target, $waste, $inputsum);
}

usdt:./src/bitcoind:coin_selection:normal_create_tx_internal
{
  $name = str(arg0);
  $success = arg1;
  $fee = (int64) arg2;
  $change_pos = (int32) arg3;
  printf("normal_create_tx_internal: wallet=%s, success=%d, fee=%d, change_pos=%d\n", $name, $success, $fee, $change_pos);
}

usdt:./src/bitcoind:coin_selection:attempting_aps_create_tx
{
  $name = str(arg0);
  printf("attempting_aps_create_tx: wallet=%s\n", $name);
}

usdt:./src/bitcoind:coin_selection:aps_create_tx_internal
{
  $name = str(arg0);
  $use_aps = arg1;
  $success = arg2;
  $fee = (int64) arg3;
  $change_pos = (int32) arg4;
  printf("aps_create_tx_internal: wallet=%s, use_aps=%d, success=%d, fee=%d, change_pos=%d\n", $name, $use_aps, $success, $fee, $change_pos);
}
```
</details>

Running e.g. the `wallet_taproot.py` functional test logs negative values for some fields (`target`, `inputsum`) where I didn't expect them. I'm not sure what's up with that. 

```
selected_coins: wallet=rpc_online, algo=bnb, target=-1436883448, waste=118372, inputsum=-1436883448
aps_create_tx_internal: wallet=rpc_online, use_aps=1, success=1, fee=131993, change_pos=-1
selected_coins: wallet=psbt_online, algo=bnb, target=-2146144343, waste=95374, inputsum=-2146144343
normal_create_tx_internal: wallet=psbt_online, success=1, fee=107994, change_pos=-1
attempting_aps_create_tx: wallet=psbt_online
selected_coins: wallet=psbt_online, algo=bnb, target=-2146144343, waste=95374, inputsum=-2146144343
aps_create_tx_internal: wallet=psbt_online, use_aps=1, success=1, fee=107994, change_pos=-1
```

achow101 commented:
> Running e.g. the `wallet_taproot.py` functional test logs negative values for some fields (`target`, `inputsum`) where I didn't expect them. I'm not sure what's up with that.

Your script uses `%d` instead of `%ld`. `%d` is for 32-bit ints, whereas `%ld` is for 64 bit.

jb55 commented:
crACK ab5af9ca7293239ffc24ea7e23159b8184543f94

0xB10C reviewed (approved):
ACK ab5af9ca7293239ffc24ea7e23159b8184543f94. Code reviewed, ran the `interface_usdt_coinselection.py` test, and tested with the above bpftrace script (updated `%d` -> `%ld` where necessary, ty achow101).

josibake commented on src/wallet/coinselection.h:
```diff
@@ -254,16 +254,17 @@ struct SelectionResult
     /** The computed waste */
     std::optional<CAmount> m_waste;
-    /** The algorithm used to produce this result */
-    SelectionAlgorithm m_algo;
```
why is `m_algo` added as private and then moved here to public? wouldn't it be better to move `m_target` and add `m_algo` in the first commit?

achow101 replied:
Indeed, I forgot to squash those separately when implementing this. Will move if I have to re-touch.

josibake reviewed (approved):
crACK https://github.com/bitcoin/bitcoin/pull/24644/commits/ab5af9ca7293239ffc24ea7e23159b8184543f94

looks good! just had one nit about commit structure. also compiled and ran the functional test without any issues

josibake commented:
tACK https://github.com/bitcoin/bitcoin/commit/ab5af9ca7293239ffc24ea7e23159b8184543f94

wasn't paying attention earlier and it was actually skipping the USDT functional tests, not running them :facepalm: . 

I got tracepoints working by running `sudo apt install systemtap-sdt-dev` (per [doc/build-unix.md](../tree/master/doc/build-unix.md)) and building `bcc` from source (https://github.com/iovisor/bcc/blob/master/INSTALL.md#ubuntu---source)

everything works as expected. i also noticed `test/functional/wallet_send.py` is failing in the CI, but i the test locally with no issues and then re-ran the whole test suite, so perhaps just a transient failure due to a timeout?

w0xlt reviewed (commented):
post-merged tACK https://github.com/bitcoin/bitcoin/pull/24644/commits/ab5af9ca7293239ffc24ea7e23159b8184543f94
Functional test ran successfully on Ubuntu 21.10 and bcc (master version).

0xB10C commented:
ref: https://github.com/bitcoin/bitcoin/pull/25003 tracing: fix coin_selection:aps_create_tx_internal calling logic

russeree commented:
Tests will fail on RHEL linux when default yum dist. of bcc is used. To remedy bcc needs to be build from bcc git master. 

Error output
https://gist.github.com/russeree/39a3f13aababe44ba8bac52224626d0b

After building and running the current master of bcc https://gist.github.com/russeree/888c52a85e9bcd69c758ff0bc91f3d13

This issue only presents on RHEL (Fedora, Centos, Redhat) Ubuntu seemed to not be affected by this using apt installed versions of bcc.
